### PR TITLE
Feature/refactor fileupload

### PR DESCRIPTION
**Pull request by atuldh**

### CODE REVIEW

File: app/uk/gov/hmrc/cbcrfrontend/connectors/FileUploadServiceConnector.scala [hunk 1]
```diff
@@ -71,5 +73,54 @@ class FileUploadServiceConnector() {
     }
   }
 
+  def extractFileUploadMessage(resp: HttpResponse): Opt[String] = {
```
Comment (MaxWorgan) [inline app/uk/gov/hmrc/cbcrfrontend/connectors/FileUploadServiceConnector.scala]: I think the Opt type alias is really confusing! Can we rename it something clearer?

File: app/uk/gov/hmrc/cbcrfrontend/connectors/FileUploadServiceConnector.scala [hunk 2]
```diff
@@ -71,5 +73,54 @@ class FileUploadServiceConnector() {
...
+      resp.status match {
+        case 200 => Right(resp.body)
+        case _ => Left(UnexpectedState("Problems uploading the file"))
+      }
+  }
+
+  def extractFileUploadResponseMessage(resp: HttpResponse): Opt[Option[FileUploadCallbackResponse]] = {
+    resp.status match {
+      case 200 => Right(Some(resp.json.as[FileUploadCallbackResponse]))
```
Comment (MaxWorgan) [inline app/uk/gov/hmrc/cbcrfrontend/connectors/FileUploadServiceConnector.scala]: I think 'as' can throw an exception - 'validate' is safer

File: app/uk/gov/hmrc/cbcrfrontend/connectors/FileUploadServiceConnector.scala [hunk 3]
```diff
@@ -71,5 +73,54 @@ class FileUploadServiceConnector() {
...
+
+        val inputStream = Source.fromString(resp.body).getCharacterStream
+        val xmlFile = File.createTempFile("xml", "xml")
+        val fos = new java.io.FileOutputStream(xmlFile)
+
+        fos.write(
+          Stream.continually(inputStream.read).takeWhile(-1 !=).map(_.toByte).toArray
+        )
+        fos.close()
+        Right(xmlFile)
```
Comment (MaxWorgan) [inline app/uk/gov/hmrc/cbcrfrontend/connectors/FileUploadServiceConnector.scala]: Did you see this?
https://www.playframework.com/documentation/2.5.x/ScalaFileUpload#direct-file-upload
Its might mean we can do away with doing our own file handling stuff...

File: app/uk/gov/hmrc/cbcrfrontend/controllers/FileUpload.scala [hunk 4]
```diff
@@ -38,105 +40,106 @@ import uk.gov.hmrc.cbcrfrontend.auth.SecuredActions
...
-  }
+        Json.obj(("fileName", fileMetadata.get.name), ("size", fileSize))
+      } else Json.obj(("fileName", "Not Found"), ("size", "Not Found"))
 
-  val fileUploadProgress = Action.async { implicit request =>
-    val envelopeId = request.flash.get("ENVELOPEID")
-    Logger.debug("Headers :"+envelopeId)
-    val hostName = FrontendAppConfig.cbcrFrontendHost
-    val assetsLocationPrefix = FrontendAppConfig.assetsPrefix
+    )
```
Comment (MaxWorgan) [inline app/uk/gov/hmrc/cbcrfrontend/controllers/FileUpload.scala]: This for comprehension has gotten a bit out of hand - I feel it's my fault :)
Maybe we can revisit this later on in the week (at least after wednesday)

File: app/uk/gov/hmrc/cbcrfrontend/model/FileMetadata.scala [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                         name: String,
+                         contentType: String,
+                         length: BigDecimal,
+                         created: String,
+                         metadata: JsValue,
+                         href: String)
+
+object FileMetadata {
+
+  implicit val fileMetadataFormat: Format[FileMetadata] = (
```
Comment (MaxWorgan) [inline app/uk/gov/hmrc/cbcrfrontend/model/FileMetadata.scala]: I don't think we actually need to declare these - ```Json.format[FileMetadata]``` is all that's needed and it'll figure it out by itself!

File: app/uk/gov/hmrc/cbcrfrontend/model/FileUploadCallbackResponse.scala [hunk 6]
```diff
@@ -16,4 +16,19 @@
...
+
+object FileUploadCallbackResponse {
+  implicit val fileUploadCallbackFormat: Format[FileUploadCallbackResponse] = (
+    (JsPath \ "envelopeId").format[String] and
+      (JsPath \ "fileId").format[String] and
+      (JsPath \ "status").format[String]
+
+    ) (FileUploadCallbackResponse.apply, unlift(FileUploadCallbackResponse.unapply))
+
+}
```
Comment (MaxWorgan) [inline app/uk/gov/hmrc/cbcrfrontend/model/FileUploadCallbackResponse.scala]: see above

File: app/uk/gov/hmrc/cbcrfrontend/controllers/FileUpload.scala (hunk 4 above)
Comment (atuldh) [inline app/uk/gov/hmrc/cbcrfrontend/controllers/FileUpload.scala]: Agree

**Pull request merged**